modules/lm-coursier/src/main/scala/lmcoursier/internal/ResolutionRun.scala

```diff
@@ -79,47 +83,112 @@ object ResolutionRun {
+                        log.info(s"Resolved ${params.projectName} dependencies")
+                    )
+                )
+              }
+            )
+        )
+    }
+
+    def retryOnFailure(maxAttempts: Int,
+                       delay: Future[Unit],
+                       retry: Int => Future[Either[ResolutionError, Resolution]],
+                       count: Int,
+                       resolutionError: Option[ResolutionError],
+                       exception: Throwable,
+                       resolution: Resolution)(implicit ec: ExecutionContext):
+    Future[Either[ResolutionError, Resolution]] = {
+
+      if (count >= maxAttempts) {
+        val ex = resolutionError.getOrElse(new ResolutionError.MaximumIterationReached(resolution))
+        log.error(s"Failed, maximum iterations ($maxAttempts) reached")
+        Future.successful(Left(ex))
+      }
+      else {
+        log.info(s"Attempt $count failed: ${resolutionError.map(_.toString).getOrElse(exception.toString)}")
+        delay.flatMap { _ =>
+          retry(count + 1)
         }
-      )
-      .withRepositories(repositories)
-      .withResolutionParams(
-        params
-          .params
-          .addForceVersion((if (isSandboxConfig) Nil else params.interProjectDependencies.map(_.moduleVersion)): _*)
-          .withForceScalaVersion(params.autoScalaLibOpt.nonEmpty)
-          .withScalaVersionOpt(params.autoScalaLibOpt.map(_._2))
-          .withTypelevel(params.params.typelevel)
-          .withRules(rules)
-      )
-      .withCache(
-        params
-          .cache
-          .withLogger(
-            params.loggerOpt.getOrElse {
-              RefreshLogger.create(
-                if (RefreshLogger.defaultFallbackMode)
-                  new FallbackRefreshDisplay()
-                else
-                  ProgressBarRefreshDisplay.create(
-                    if (printOptionalMessage) log.info(initialMessage),
-                    if (printOptionalMessage || verbosityLevel >= 2)
-                      log.info(s"Resolved ${params.projectName} dependencies")
-                  )
-              )
+      }
+    }
+
+    val finalTask: Either[ResolutionError, Resolution] =
+      params.retry match {
+        case (period, maxAttempts) =>
+          val scheduler = Executors.newSingleThreadScheduledExecutor(
+            new ThreadFactory {
+              val defaultThreadFactory: ThreadFactory = Executors.defaultThreadFactory()
+
+              def newThread(r: Runnable): Thread = {
+                val t = defaultThreadFactory.newThread(r)
+                t.setDaemon(true)
+                t.setName("retry-handler")
+                t
+              }
             }
           )
-      )
-      .either() match {
-        case Left(err) if params.missingOk => Right(err.resolution)
-        case others => others
+          implicit val ec: ExecutionContext = resolveTask.cache.ec
+
+          val delay = Task.completeAfter(scheduler, period).future()
```

Comment:
Does this `val` need to be a `def`? Wouldn't assigning a Future basically start the clock?
Also maybe it should be an exponential backoff, if not linear backoff `count * some-time`.

Reply:
fixed